### PR TITLE
Improve `AffineDecomposition`

### LINKED ISSUE #40: Generalize `coefficient_map` from `AffineDecomposition`s
There are affine decompositions, such as the magnetization $2L^{-1} \sum_iS_i^z$, that don't have a parameter-dependent coefficients. In these cases it would be sensible to have a `coefficient_map` that is just a vector of coefficients instead of a function returning that vector.

Then it would be possible to dispatch `compress` on `AffineDecomposition{T, <:AbstractArray{<:Number}}` which would return the (constant) compressed matrix directly instead of another `AffineDecomposition`.

We would then also avoid the inconvenient `m_reduced = m([])` definitions in the docs examples.

### PR DESCRIPTION
**Pull request by pbrehmer**
The type restriction on the `coefficient_map` should be lifted in order to represent affine decompositions with constant coefficients. Also, as it turns out, this seems necessary to allow for a working custom serialization of `AffineDecomposition`s.

@mfherbst Do you think we should implement a `compress(::AffineDecomposition{<:AbstractArray,<:AbstractArray}, ...)` for affine decompositions with constant coefficients that returns the compressed matrix (as we shortly discussed with regards to `m_reduced` in the docs examples)?

It could be that this makes things more confusing, since `compress` would then return 3 objects (the matrix, the compressed `AffineDecomposition` and a `AffineDecomposition` with only the matrix elements, which is needed for later truncation).

- closes #40

### CODE REVIEW

File: examples/multi_ad.jl [hunk 1]
```diff
@@ -84,7 +84,7 @@ terms = map(idx -> to_global(σz, L, first(idx.I)) * to_global(σz, L, last(idx.
 # Correspondingly, the coefficient function now has to map one ``k`` value to a matrix of
 # coefficients of the same size as the `terms` matrix:
 
-coefficient_map = k -> map(idx -> cis(-(first(idx.I) - last(idx.I)) * k) / L,
+coefficients = k -> map(idx -> cis(-(first(idx.I) - last(idx.I)) * k) / L,
                            CartesianIndices((1:L, 1:L)));
```
Comment (mfherbst) [inline examples/multi_ad.jl:88]: alignment

File: src/affine_decomposition.jl [hunk 2]
```diff
@@ -6,21 +6,24 @@ O(\\bm{\\mu}) = \\sum_{r=1}^R \\alpha_r(\\bm{\\mu})\\, O_r
...
+the parameter vector, `coefficients` can also just be an `AbstractArray` of the same size
+as `terms`.
 
 Note that ``r = (r_1, \\dots, r_d)`` generally is a multi-index and as such `terms`
-can be a ``d``-dimensional array. Correspondingly, `coefficient_map` maps parameter points
+can be a ``d``-dimensional array. Correspondingly, `coefficients` maps parameter points
 ``\\bm{\\mu}`` to an `size(terms)` array.
 """
-struct AffineDecomposition{T<:AbstractArray,F<:Function}
+struct AffineDecomposition{T<:AbstractArray,F}
```
Comment (mfherbst) [inline src/affine_decomposition.jl:19]: Maybe `F` is not the best type symbol then. Maybe `C` ?

**Review by mfherbst — COMMENTED** [review]
Nits

**Comment by mfherbst** [comment]
> @mfherbst Do you think we should implement a compress(::AffineDecomposition{<:AbstractArray,<:AbstractArray}, ...) for affine decompositions with constant coefficients that returns the compressed matrix (as we shortly discussed with regards to m_reduced in the docs examples)?

> It could be that this makes things more confusing, since compress would then return 3 objects (the matrix, the compressed AffineDecomposition and a AffineDecomposition with only the matrix elements, which is needed for later truncation).

Generally I would indeed discourage this. On the other hand if you use named tuples this is not a problem, since then one can either choose to use this feature or not. In my opinion the deciding factor is whether this either improves convenience (in the sense that one saves a lot of repetitive code) or it improves performance. I don't think it is the latter, but it might be the former. What do you think?

**Comment by pbrehmer** [comment]
> > @mfherbst Do you think we should implement a compress(::AffineDecomposition{<:AbstractArray,<:AbstractArray}, ...) for affine decompositions with constant coefficients that returns the compressed matrix (as we shortly discussed with regards to m_reduced in the docs examples)?
> 
> > It could be that this makes things more confusing, since compress would then return 3 objects (the matrix, the compressed AffineDecomposition and a AffineDecomposition with only the matrix elements, which is needed for later truncation).
> 
> Generally I would indeed discourage this. On the other hand if you use named tuples this is not a problem, since then one can either choose to use this feature or not. In my opinion the deciding factor is whether this either improves convenience (in the sense that one saves a lot of repetitive code) or it improves performance. I don't think it is the latter, but it might be the former. What do you think?

My Feeling is that including this option does not avoid that much code repetition. Also, I think this would be confusing to describe in the Docs examples, even more so than explaining that some `AffineDecomposition`s have constant coefficients such that the explicit sum can be obtained by calling just `(::AffineDecomposition)()`. So for now, I would prefer to exclude the matrix-return option.

**Comment by mfherbst** [comment]
Ok, then let's don't. Just rebase the code and the we merge.

**Comment by pbrehmer** [comment]
I would have one more thing we might add in this PR on the topic of improving `AffineDecomposition`s. So in practice one needs a parallelized version of `compress`, e.g. when computing the compressed terms for a structure factor — they can be evaluated all in parallel since there is no data dependency between the terms.

Now in my private RB applications repository, I have already implemented such a method for affine decompositions with matrix-like terms:

```julia
function compress(ad::AffineDecomposition{<:Matrix}, basis::RBasis; symmetric_terms=false)
    matrixel = similar(ad.terms, Matrix{Number})
    is_nonredundant(a, b) = (size(ad.terms, 1) > size(ad.terms, 2)) ? ≥(a, b) : ≤(a, b)
    Threads.@threads for idx in CartesianIndices(matrixel)
        if is_nonredundant(first(idx.I), last(idx.I))  # Redundant upper/lower triangle
            matrixel[idx] = compress(ad.terms[idx], basis.snapshots)
        end
    end  
    
    Threads.@threads for idx in findall(x -> !is_nonredundant(first(x.I), last(x.I)), matrixel)
        if symmetric_terms
            matrixel[idx] = matrixel[last(idx.I), first(idx.I)]
        else
            matrixel[idx] = compress(ad.terms[idx], basis.snapshots)
        end
    end  # Use "symmetry" to fill redundant elements
    matrixel = promote_type.(matrixel)  # Promote to common floating-point type
    rbterms = map(m -> basis.vectors' * m * basis.vectors, matrixel)
    
    (; rb=AffineDecomposition(rbterms, ad.coefficients),
     raw=AffineDecomposition(matrixel, ad.coefficients))
end
```

Of course this could be generalized to any type of terms, but this particular method is especially needed at the moment. Unfortunately, this still uses the `promote_type` but I don't know how to avoid this at the moment.

Note that this also touches the issue #35, which we would have to rethink anyway in order to implement a parallelized `compress`.

**Comment by pbrehmer** [comment]
@mfherbst I think I'll leave this `compress` feature for now and just use the one I implemented in my (still private) applications repository. I'll just resolve the conflicts and then we can merge.

**Pull request merged**